File: LinkedList/RemoveDuplicatesfromSortedListII.java
/*
Given a sorted linked list, delete all nodes that have duplicate numbers, 
leaving only distinct numbers from the original list.

For example,
Given 1->2->3->3->4->4->5, return 1->2->5.
Given 1->1->1->2->3, return 2->3.
*/

/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) { val = x; }
 * }
 */
public class Solution {
    public ListNode deleteDuplicates(ListNode head) {
        if(head == null) return head;
        ListNode dummy = new ListNode(0);
        dummy.next = head;
        ListNode pre = dummy;
        ListNode cur = head;
        while(cur != null){
            while(cur.next != null && cur.next.val == cur.val){
                cur = cur.next;
            }
            if(pre.next == cur) {
                pre = pre.next;
            }
            else pre.next = cur.next;
            cur = cur.next;
        }
        return dummy.next;
    }
}

File: LinkedList/RemoveDuplicatesfromSortedListII.js
/*
Given a sorted linked list, delete all nodes that have duplicate numbers, 
leaving only distinct numbers from the original list.

For example,
Given 1->2->3->3->4->4->5, return 1->2->5.
Given 1->1->1->2->3, return 2->3.
*/

/**
 * Definition for singly-linked list.
 * function ListNode(val) {
 *     this.val = val;
 *     this.next = null;
 * }
 */
/**
 * @param {ListNode} head
 * @return {ListNode}
 */
var deleteDuplicates = function(head) {
    if(head === null) return head;
    var dummy = new ListNode(0);
    dummy.next = head;
    var p = dummy;
    while(head !== null){
        while(head.next !== null && head.next.val == head.val){
            head = head.next;
        }
        if(p.next == head){
            p = p.next;
        }
        else p.next = head.next;
        head = head.next;
    }
    return dummy.next;
};